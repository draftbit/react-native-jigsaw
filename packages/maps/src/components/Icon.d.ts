import * as React from "react";
import { ViewProps, StyleProp, ImageStyle } from "react-native";
declare type Props = {
  name:
    | string
    | number
    | {
        uri: string;
      };
  color?: string;
  size: number;
  style?: StyleProp<ImageStyle>;
} & ViewProps;
declare const Icon: React.FC<Props>;
export default Icon;
