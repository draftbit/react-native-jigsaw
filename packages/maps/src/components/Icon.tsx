import * as React from "react";
import {
  Image,
  View,
  StyleSheet,
  ViewProps,
  StyleProp,
  ImageStyle,
  Platform,
} from "react-native";

// This must use require to work in both web as a published project and in Snack
const VectorIcons = require("@expo/vector-icons");

type Props = {
  name: string | number | { uri: string };
  color?: string;
  size: number;
  style?: StyleProp<ImageStyle>;
} & ViewProps;

const Icon: React.FC<Props> = ({ name, color, size, style, ...rest }) => {
  if (!name) return null;

  let iconSet = "MaterialIcons";
  if (typeof name === "string" && name.indexOf("/") !== -1) {
    [iconSet, name] = name.split("/");
  }

  if (typeof name === "string") {
    const IconSet = VectorIcons[iconSet];

    return (
      <View style={[styles.container, { width: size, height: size }, style]}>
        <IconSet {...rest} name={name} color={color} size={size} />
      </View>
    );
  } else if (
    (typeof name === "object" &&
      name !== null &&
      Object.prototype.hasOwnProperty.call(name, "uri") &&
      typeof name.uri === "string") ||
    typeof name === "number"
  ) {
    return (
      <Image
        {...rest}
        source={name}
        style={[
          {
            width: size,
            height: size,
            tintColor: color,
          },
          style,
        ]}
      />
    );
  }

  return (
    <View
      {...rest}
      style={[
        {
          width: size,
          height: size,
        },
        styles.container,
        style,
      ]}
    >
      {name}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    justifyContent: "center",
    overflow: "hidden",
    ...Platform.select({
      web: {
        cursor: "pointer",
        userSelect: "none",
      },
    }),
  },
});

export default Icon;
