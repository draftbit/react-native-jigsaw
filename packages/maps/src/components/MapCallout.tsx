import * as React from "react";
import { Callout } from "./ReactNativeMaps";
import { MapCalloutProps } from "@draftbit/types";

const MapCallout: React.FC<MapCalloutProps> = ({
  onPress,
  showTooltip,
  children,
}) => {
  if (!Callout) {
    return null;
  }

  return (
    <Callout tooltip={!showTooltip} onPress={onPress}>
      {children}
    </Callout>
  );
};

export default MapCallout;
