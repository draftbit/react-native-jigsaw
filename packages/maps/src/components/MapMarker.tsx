import * as React from "react";
import { Marker } from "./ReactNativeMaps";
import { MapMarkerProps } from "@draftbit/types";

const MapMarker: React.FC<React.PropsWithChildren<MapMarkerProps>> = ({
  latitude,
  longitude,
  title,
  description,
  pinColor,
  flat,
  style,
  children,
}) => {
  if (!Marker) {
    return null;
  }

  return (
    <Marker
      coordinate={{
        latitude,
        longitude,
      }}
      title={title != null ? String(title) : undefined}
      description={description != null ? String(description) : undefined}
      flat={flat}
      pinColor={pinColor}
      style={style}
    >
      {children}
    </Marker>
  );
};

export default MapMarker;
