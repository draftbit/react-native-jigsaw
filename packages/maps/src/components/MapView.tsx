import * as React from "react";
import NativeMapView, { Marker } from "./ReactNativeMaps";
import { MapViewProps } from "@draftbit/types";

// Dynamically import from ./ReactNativeMaps so that we don't
// require react-native-maps unless we're in native.

// Approximates a conversion of zoom level to altitude,
// since different platforms require different values.
// https://stackoverflow.com/a/37142662
function zoomToAltitude(zoom: number) {
  const A = 40487.57;
  const B = 0.00007096758;
  const C = 91610.74;
  const D = -40467.74;

  return C * Math.pow((A - D) / (zoom - D) - 1, 1 / B);
}

class MapView extends React.Component<MapViewProps> {
  private mapRef: React.RefObject<any>;
  constructor(props: MapViewProps) {
    super(props);
    this.state = {};
    this.mapRef = React.createRef();
  }

  componentDidUpdate(prevProps: MapViewProps) {
    if (
      prevProps.latitude != null &&
      prevProps.longitude != null &&
      this.props.latitude != null &&
      this.props.longitude != null &&
      (prevProps.latitude !== this.props.latitude ||
        prevProps.longitude !== this.props.longitude)
    ) {
      this.animateToLocation({
        latitude: this.props.latitude,
        longitude: this.props.longitude,
        zoom: this.props.zoom,
      });
    }
  }

  animateToLocation({
    latitude,
    longitude,
    zoom,
  }: {
    latitude: number;
    longitude: number;
    zoom?: number;
  }) {
    const args: {
      center: { latitude: number; longitude: number };
      altitude?: number;
      zoom?: number;
    } = {
      center: {
        latitude,
        longitude,
      },
    };

    if (zoom) {
      args.altitude = zoomToAltitude(zoom || 1);
      args.zoom = zoom;
    }

    this.mapRef.current.animateCamera(args);
  }

  render() {
    const {
      provider,
      latitude,
      longitude,
      zoom,
      showsCompass = false,
      rotateEnabled = true,
      zoomEnabled = true,
      loadingEnabled = true,
      scrollEnabled = true,
      loadingBackgroundColor,
      loadingIndicatorColor,
      mapType = "standard",
      showsUserLocation,
      followsUserLocation,
      showsPointsOfInterest,
      style,
      children,
    } = this.props;

    if (!NativeMapView || !Marker) {
      return null;
    }

    const camera = {
      altitude: zoomToAltitude(zoom || 1),
      heading: 0,
      pitch: 0,
      zoom,
      center: {
        latitude,
        longitude,
      },
    };

    return (
      <NativeMapView
        ref={this.mapRef}
        provider={provider}
        mapType={mapType}
        showsCompass={showsCompass}
        rotateEnabled={rotateEnabled}
        zoomEnabled={zoomEnabled}
        camera={camera}
        loadingEnabled={loadingEnabled}
        scrollEnabled={scrollEnabled}
        showsUserLocation={showsUserLocation}
        followsUserLocation={followsUserLocation}
        showsPointsOfInterest={showsPointsOfInterest}
        loadingBackgroundColor={loadingBackgroundColor}
        loadingIndicatorColor={loadingIndicatorColor}
        style={style}
      >
        {children}
      </NativeMapView>
    );
  }
}

export default MapView;
