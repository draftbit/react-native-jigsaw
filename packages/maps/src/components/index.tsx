export { default as MapCallout } from "./MapCallout";
export { default as MapMarker } from "./MapMarker";
export { default as MapView } from "./MapView";
