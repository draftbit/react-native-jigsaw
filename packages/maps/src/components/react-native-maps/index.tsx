export { default as default } from "./MapView";
export { Callout } from "./Callout";
export { Marker } from "./Marker";
export { Circle } from "./Circle";
