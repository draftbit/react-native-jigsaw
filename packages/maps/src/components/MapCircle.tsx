import * as React from "react";
import { Platform } from "react-native";
import { Circle as MapCircleComponent } from "./react-native-maps";
import type { MapCircleProps as MapCircleComponentProps } from "react-native-maps";
import { withTheme, DefaultTheme } from "@draftbit/ui";
import Color from "color";

export interface MapCircleProps
  extends Omit<MapCircleComponentProps, "center"> {
  latitude: number;
  longitude: number;
  theme: typeof DefaultTheme;
}

const MapCircle: React.FC<React.PropsWithChildren<MapCircleProps>> = ({
  theme,
  latitude,
  longitude,
  radius = 2000,
  fillColor: fillColorProp = theme.colors.primary,
  strokeColor = theme.colors.primary,
  ...rest
}) => {
  const parsedColor = Color(fillColorProp);

  let fillColor;
  if (parsedColor.alpha() === 0) {
    fillColor = "transparent";
  } else if (Platform.OS !== "web") {
    // Web maps by default uses a lower opacity for the circle, native needs this extra step
    fillColor = parsedColor.alpha(0.3).rgb().string();
  } else {
    fillColor = fillColorProp;
  }

  return (
    <MapCircleComponent
      center={{
        latitude,
        longitude,
      }}
      radius={radius}
      fillColor={fillColor}
      strokeColor={strokeColor}
      {...rest}
    />
  );
};

export default withTheme(MapCircle);
