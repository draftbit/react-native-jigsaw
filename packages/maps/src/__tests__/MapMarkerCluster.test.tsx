import React from "react";
import { Text } from "react-native";
import { render, screen } from "@testing-library/react-native";
import MapMarker from "../components/MapMarker";
import {
  MapMarkerCluster,
  MapMarkerClusterView,
} from "../components/marker-cluster";
import { Marker as MapMarkerComponent } from "../components/react-native-maps";

jest.mock("../components/react-native-maps", () => {
  const React = require("react");

  class MapView extends React.Component {
    render(): React.ReactNode {
      return <>{this.props.children}</>;
    }
  }

  const Marker = (props: any) => {
    return <>{props.children}</>;
  };

  return {
    __esModule: true,
    default: MapView,
    Marker,
  };
});

jest.mock("@teovilla/react-native-web-maps", () => {
  const React = require("react");

  class MarkerClusterer extends React.Component {
    render(): React.ReactNode {
      return (
        <>
          {this.props.renderCluster?.({
            pointCount: React.Children.toArray(this.props.children).length,
            coordinate: { latitude: 0, longitude: 0 },
            expansionZoom: 4,
          })}
          {this.props.children}
        </>
      );
    }
  }

  return {
    MarkerClusterer,
  };
});

beforeEach(() => {
  jest.clearAllMocks();
});

describe("MapMarkerCluster tests", () => {
  test("should render markers provided as children", () => {
    render(
      <MapMarkerCluster>
        <MapMarker latitude={41.741895} longitude={-73.989308} />
        <MapMarker latitude={42.741895} longitude={-73.989308} />
        <MapMarker latitude={43.741895} longitude={-73.989308} />
        <MapMarker latitude={44.741895} longitude={-73.989308} />
      </MapMarkerCluster>
    );

    const renderedMarkers = screen.UNSAFE_queryAllByType(MapMarkerComponent);

    expect(renderedMarkers.length).toBe(4 + 1); //4 clustered markers, cluster itself is also rendered as it's own marker (reasoning behind the +1)
  });

  test("should render default cluster view when none provided", () => {
    render(
      <MapMarkerCluster>
        <MapMarker latitude={41.741895} longitude={-73.989308} />
      </MapMarkerCluster>
    );

    const defaultClusterView = screen.queryByTestId(
      "default-map-marker-cluster-view"
    );

    expect(defaultClusterView).toBeTruthy();
  });

  test("should render custom cluster view when provided", () => {
    render(
      <MapMarkerCluster>
        <MapMarkerClusterView
          renderItem={({ markerCount }) => (
            <Text testID="custom-cluster-view">{markerCount}</Text>
          )}
        />
        <MapMarker latitude={41.741895} longitude={-73.989308} />
        <MapMarker latitude={42.741895} longitude={-73.989308} />
        <MapMarker latitude={43.741895} longitude={-73.989308} />
      </MapMarkerCluster>
    );

    const customClusterView = screen.queryByTestId("custom-cluster-view");

    expect(customClusterView).toBeTruthy();
    expect(customClusterView?.props).toMatchInlineSnapshot(`
      {
        "children": 3,
        "testID": "custom-cluster-view",
      }
    `);
  });
});
