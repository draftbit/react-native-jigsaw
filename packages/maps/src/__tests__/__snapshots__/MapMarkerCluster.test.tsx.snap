// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapMarkerCluster tests should render custom cluster view when provided 1`] = `
{
  "children": 3,
  "testID": "custom-cluster-view",
}
`;
